ClosedXML/Excel/XLWorkbook_Save.cs

```diff
@@ -2357,19 +2357,35 @@ private static void GenerateTableDefinitionPartContent(TableDefinitionPart table
             var pageFields = new PageFields { Count = (uint)pt.ReportFilters.Count() };
             var pivotFields = new PivotFields { Count = Convert.ToUInt32(pt.SourceRange.ColumnCount()) };
 
-            foreach (var xlpf in pt
-                .Fields
-                .Cast<XLPivotField>()
-                .OrderBy(f => pt.RowLabels.Any(p => p.SourceName == f.SourceName)
-                              ? pt.RowLabels.IndexOf(f)
-                              : Int32.MaxValue))
+            var orderedPageFields = new SortedDictionary<int, PageField>();
+            var orderedColumnLabels = new SortedDictionary<int, Field>();
+            var orderedRowLabels = new SortedDictionary<int, Field>();
+
+            // Add value fields first
+            if (pt.Values.Any())
+            {
+                if (pt.RowLabels.Contains(XLConstants.PivotTableValuesSentinalLabel))
+                {
+                    var f = pt.RowLabels.First(f1 => f1.SourceName == XLConstants.PivotTableValuesSentinalLabel);
+                    orderedRowLabels.Add(pt.RowLabels.IndexOf(f), new Field { Index = -2 });
+                    pivotTableDefinition.DataOnRows = true;
+                }
+                else if (pt.ColumnLabels.Contains(XLConstants.PivotTableValuesSentinalLabel))
+                {
+                    var f = pt.ColumnLabels.First(f1 => f1.SourceName == XLConstants.PivotTableValuesSentinalLabel);
+                    orderedColumnLabels.Add(pt.ColumnLabels.IndexOf(f), new Field { Index = -2 });
+                }
+            }
+
+            foreach (var xlpf in pt.Fields)
```

Comment:
This loop looks suboptimal. We enumerate through a collection which gives us the complexity O(n), and inside the loop, we call `pt.RowLabels.Contains()`, `pt.RowLabels.IndexOf()`, etc. - each of them has the complexity of O(n) too. Thus we get an overall complexity O(n²) on a relatively simple task. It's not the problem while the pivot tables have a reasonable size but we can have performance issues here in future.

Reply:
Yes, in fact, while pivot tables are still kind of WIP I'm not worrying too much. Once pivot tables are considered stable, I want to do a full refactoring.